synapse/axon.py

```diff
@@ -1009,6 +1048,72 @@ async def readlines(self, sha256):
         if remain:
             yield remain
 
+    @staticmethod
+    def _getCsvReader(lines, dialect, **fmtparams) -> csv.reader:
+        dialect = dialect.lower()
+        if dialect not in csv.list_dialects():
+            raise s_exc.BadArg(mesg=f'Invalid CSV dialect, use one of {csv.list_dialects()}')
+
+        try:
+            reader = csv.reader(lines, dialect, **fmtparams)
+        except TypeError as e:
+            if 'invalid keyword' in str(e):
+                raise s_exc.BadArg(mesg=f'Invalid csv format parameter: {str(e)}') from None
+            else:  # pragma: no cover
+                raise
+        else:
+            return reader
+
+    @contextlib.asynccontextmanager
+    async def open(self, sha256, mode='r'):
+
+        link, fobj = await s_link.linkfile(mode=mode)
+
+        async def fill():
+            async for byts in self.get(sha256):
+                await link.send(byts)
+            await link.fini()
+
+        self.schedCoro(fill())
+
+        yield fobj
+
+        fobj.close()
+        await link.fini()
+
+    def _csvrows(self, fobj, queue, sha256, dialect, **fmtparams):
+        try:
+            reader = self._getCsvReader(fobj, dialect, **fmtparams)
+            for row in reader:
+                queue.put(row)
+        except StopIteration:
+            pass
+        except (UnicodeDecodeError, csv.Error) as e:
+            raise s_exc.BadDataValu(mesg=f'Error processing csv row: {e}', sha256=sha256) from None
+        except Exception:
+            logger.exception('Error processing csv.')
+            raise
+        finally:
+            queue.put(None)
+
+    async def csvrows(self, sha256, dialect='excel', **fmtparams):
```

Comment:
axon APIs take binary sha256 ( caller should do uhex )

Reply:
readlines and jsonlines broke that convention :(